File: strands-ts/src/context-manager/context-manager.ts
/**
 * ContextManager: first-class agent component for strategy-driven context management.
 *
 * On overflow, runs the strategy pipeline (including an emergency truncation as the final step).
 */

import type { Plugin } from '../plugins/plugin.js'
import type { Tool } from '../tools/tool.js'
import type { LocalAgent } from '../types/agent.js'
import { AfterModelCallEvent, BeforeModelCallEvent, MessageAddedEvent } from '../hooks/events.js'
import { ContextWindowOverflowError } from '../errors.js'
import { InMemoryStorage } from '../storage/in-memory-storage.js'
import { EPHEMERAL } from '../storage/storage.js'
import type { Storage } from '../storage/storage.js'
import { logger } from '../logging/logger.js'
import type { ContextManagerConfig, ContextStrategy, ContextState } from './types.js'
import { EmergencyTruncateStrategy, Offload } from './strategies/offload/index.js'
import { resolveStrategies } from './presets.js'
import { Stash } from './stash.js'
import { createRetrievalTool, trackRetrievalToolUseIds } from './retrieval-tool.js'

const AUTO_TRUNCATE_THRESHOLD = 1_500
const AGENTIC_TRUNCATE_THRESHOLD = 8_000
const TRUNCATE_PREVIEW_TOKENS = 750
const AUTO_SUMMARIZE_UTILIZATION = 0.85

/** @internal */
export const CONTEXT_MANAGER_PRESETS = ['auto', 'agentic'] as const

/** @internal */
export type ContextManagerPreset = (typeof CONTEXT_MANAGER_PRESETS)[number]

/**
 * Supported values for the `contextManager` parameter.
 *
 * - `"auto"`: Managed context with proactive compression + offloading.
 * - `"agentic"`: Model-driven context management via injected tools.
 * - `ContextManagerConfig`: Custom strategy pipeline and stash configuration.
 * - `false`: Explicitly disable all context management (no compression, no offloading).
 */
export type ContextManagerStrategy = ContextManagerPreset | ContextManagerConfig | false

/**
 * Manages context reduction for an agent's conversation.
 *
 * On context overflow, runs the strategy pipeline (offload, summarize, emergency truncate).
 * The emergency truncation is always appended as the final strategy — it recomputes
 * utilization and only fires if the window is still overflowing after user strategies.
 *
 * Configured through the Agent's `contextManager` parameter — pass a preset
 * string (`'auto'`, `'agentic'`) or a `ContextManagerConfig`; the Agent
 * constructs and registers the manager. When present, it owns overflow
 * recovery and proactive compression — no separate ConversationManager is needed.
 *
 * @experimental
 */
export class ContextManager implements Plugin {
  readonly name = 'strands:context-manager'

  private readonly _strategies: ContextStrategy[]
  private readonly _stashStorage: Storage | false | undefined
  private readonly _enableRetrievalTool: boolean
  private readonly _retrievalToolUseIds = new Set<string>()
  private _stash: Stash | undefined
  private _stashIsDurable = false
  private _retrievalTool: Tool | undefined

  constructor(config?: ContextManagerConfig) {
    const userStrategies = config?.strategies
      ? resolveStrategies(config.strategies)
      : [
          Offload.truncate('toolResults', { previewTokens: TRUNCATE_PREVIEW_TOKENS }).when({
            threshold: AUTO_TRUNCATE_THRESHOLD,
          }),
          Offload.summarize('*').when({ utilization: AUTO_SUMMARIZE_UTILIZATION, preserveRecent: 4 }),
        ]
    this._strategies = [...userStrategies, new EmergencyTruncateStrategy()]
    const stashConfig = config?.stash
    const stashObj = typeof stashConfig === 'object' ? stashConfig : undefined
    this._stashStorage = stashConfig === false ? false : stashObj?.storage
    this._enableRetrievalTool = stashConfig !== false && stashObj?.retrievalTool !== false
  }

  /**
   * Resolves a `ContextManagerStrategy` value into a `ContextManager` instance.
   *
   * @param strategy - A preset string, config object, false, or undefined
   * @returns A ContextManager for preset strings and configs; undefined for false/undefined
   */
  static from(strategy: ContextManagerStrategy | undefined): ContextManager | undefined {
    if (strategy === false || strategy === undefined) return undefined
    if (strategy === 'auto') {
      return new ContextManager()
    }
    if (strategy === 'agentic') {
      return new ContextManager({
        strategies: [
          Offload.truncate('toolResults', { previewTokens: TRUNCATE_PREVIEW_TOKENS }).when({
            threshold: AGENTIC_TRUNCATE_THRESHOLD,
          }),
          Offload.summarize('*').when({ utilization: 1, preserveRecent: 4 }),
        ],
      })
    }
    if (typeof strategy === 'string') {
      throw new Error(
        `Unknown contextManager preset: "${strategy}". Valid presets: ${CONTEXT_MANAGER_PRESETS.map((s) => `"${s}"`).join(', ')}`
      )
    }
    return new ContextManager(strategy)
  }

  getTools(): Tool[] {
    if (!this._enableRetrievalTool) return []
    if (!this._stash) return []
    if (!this._retrievalTool) {
      this._retrievalTool = createRetrievalTool(this._stash)
    }
    return [this._retrievalTool]
  }

  async initAgent(agent: LocalAgent): Promise<void> {
    if (this._stashStorage !== false) {
      const storage = this._stashStorage ?? agent.storage ?? new InMemoryStorage()
      this._stashIsDurable = !(EPHEMERAL in storage)
      this._stash = new Stash(storage, agent.sessionId, agent.id)
    }

    if (this._stash) {
      const stash = this._stash
      const skipSet = this._retrievalToolUseIds

      for (const message of agent.messages) {
        trackRetrievalToolUseIds(message, skipSet)
        await stash.storeMessage(message, skipSet)
      }

      agent.addHook(MessageAddedEvent, async (event) => {
        trackRetrievalToolUseIds(event.message, skipSet)
        await stash.storeMessage(event.message, skipSet)
      })
    }

    for (const strategy of this._strategies) {
      strategy.init?.(agent, this._stash)
    }

    agent.addHook(BeforeModelCallEvent, async (event) => {
      await this._runStrategies(event.agent, event.projectedInputTokens)
    })

    // Assumes sequential invocations on this agent (no concurrent calls)
    let overflowRetries = 0
    agent.addHook(AfterModelCallEvent, async (event) => {
      if (!(event.error instanceof ContextWindowOverflowError)) {
        overflowRetries = 0
        return
      }

      if (overflowRetries >= 3) {
        logger.warn(`agentId=<${event.agent.id}> | overflow retry limit reached, giving up`)
        overflowRetries = 0
        return
      }

      const acted = await this._runStrategies(event.agent, undefined, true)
      if (!acted) {
        logger.warn(`agentId=<${event.agent.id}> | no strategy made progress, skipping retry`)
        return
      }

      overflowRetries++
      event.retry = true
    })
  }

  /**
   * The L1 stash instance, if stash is enabled and the agent has been initialized.
   *
   * @returns The stash, or undefined if stash is disabled or initAgent has not run
   */
  get stash(): Stash | undefined {
    return this._stash
  }

  /**
   * Whether the stash is backed by durable storage that survives process restarts.
   * When true, stash data does not need to be embedded in session snapshots.
   *
   * @returns true if the stash storage is not ephemeral
   */
  get stashIsDurable(): boolean {
    return this._stashIsDurable
  }

  private async _runStrategies(
    agent: LocalAgent,
    precomputedInputTokens?: number,
    overflow?: boolean
  ): Promise<boolean> {
    const messages = agent.messages
    const inputTokens = precomputedInputTokens ?? (await agent.model.countTokens(messages))

    const strategyContext: ContextState = {
      messages,
      agent,
      utilization: agent.model.estimateUtilization(inputTokens),
      ...(overflow ? { overflow: true } : {}),
      ...(this._stash ? { stash: this._stash } : {}),
    }

    let anyActed = false
    for (const strategy of this._strategies) {
      try {
        const acted = await strategy.apply(strategyContext)
        if (acted) {
          anyActed = true
          strategyContext.utilization = agent.model.estimateUtilization(await agent.model.countTokens(messages))
          if (strategyContext.overflow && strategyContext.utilization < 1.0) {
            strategyContext.overflow = false
          }
          logger.debug(`strategy=<${strategy.name}>, agentId=<${agent.id}> | strategy applied`)
        }
      } catch (error) {
        logger.warn(
          `strategy=<${strategy.name}>, agentId=<${agent.id}>, error=<${error}> | strategy failed, continuing`
        )
      }
    }
    return anyActed
  }
}

File: strands-py/src/strands/_context_manager/context_manager.py
"""ContextManager: first-class agent component for strategy-driven context management.

On overflow, runs the strategy pipeline (including an emergency truncation as the final step).
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from ..hooks.events import AfterModelCallEvent, BeforeModelCallEvent
from ..plugins.plugin import Plugin
from ..types.exceptions import ContextWindowOverflowException
from .strategies.offload import Offload
from .strategies.offload.base import EmergencyTruncateStrategy
from .types import ContextState, ContextStrategy

if TYPE_CHECKING:
    from ..agent.agent import Agent

logger = logging.getLogger(__name__)


class ContextManager(Plugin):
    """Manages context reduction for an agent's conversation.

    On context overflow, runs the strategy pipeline (offload, summarize, emergency truncate).
    The emergency truncation is always appended as the final strategy — it recomputes
    utilization and only fires if the window is still overflowing after user strategies.

    Pass via the ``context_manager`` parameter on the Agent constructor. When present,
    it owns overflow recovery — no separate ConversationManager is needed.
    """

    @property
    def name(self) -> str:
        """Plugin name."""
        return "strands:context-manager"

    def __init__(self, *, strategies: list[ContextStrategy] | None = None) -> None:
        """Initialize with an optional ordered list of strategies (defaults provided)."""
        super().__init__()
        self._strategies: list[ContextStrategy] = [
            *(
                strategies
                if strategies is not None
                else [
                    Offload.truncate("tool_results").when(threshold=2500),
                    Offload.summarize("*").when(threshold=1000, utilization=0.85),
                ]
            ),
            EmergencyTruncateStrategy(),
        ]

    def init_agent(self, agent: Agent) -> None:
        """Register strategy hooks for proactive compression and overflow recovery."""
        for strategy in self._strategies:
            init = getattr(strategy, "init", None)
            if init is not None:
                init(agent)

        async def _on_before_model_call(event: BeforeModelCallEvent) -> None:
            await self._run_strategies(event.agent, event.projected_input_tokens)

        agent.hooks.add_callback(BeforeModelCallEvent, _on_before_model_call)

        overflow_retries = 0

        async def _on_after_model_call(event: AfterModelCallEvent) -> None:
            nonlocal overflow_retries

            if not isinstance(event.exception, ContextWindowOverflowException):
                overflow_retries = 0
                return

            if overflow_retries >= 3:
                logger.warning("agent_id=<%s> | overflow retry limit reached, giving up", event.agent.agent_id)
                overflow_retries = 0
                return

            acted = await self._run_strategies(event.agent)
            if not acted:
                logger.warning("agent_id=<%s> | no strategy made progress, skipping retry", event.agent.agent_id)
                return

            overflow_retries += 1
            event.retry = True

        agent.hooks.add_callback(AfterModelCallEvent, _on_after_model_call)

    async def _run_strategies(self, agent: Agent, precomputed_input_tokens: int | None = None) -> bool:
        """Run the strategy pipeline, recomputing utilization after each acting strategy."""
        messages = agent.messages
        if precomputed_input_tokens is not None:
            input_tokens = precomputed_input_tokens
        else:
            input_tokens = await agent.model.count_tokens(messages)

        context = ContextState(
            messages=messages,
            agent=agent,
            utilization=agent.model.estimate_utilization(input_tokens),
        )

        any_acted = False
        for strategy in self._strategies:
            try:
                acted = await strategy.apply(context)
                if acted:
                    any_acted = True
                    new_tokens = await agent.model.count_tokens(messages)
                    context.utilization = agent.model.estimate_utilization(new_tokens)
                    logger.debug("strategy=<%s>, agent_id=<%s> | strategy applied", strategy.name, agent.agent_id)
            except Exception:
                logger.warning(
                    "strategy=<%s>, agent_id=<%s> | strategy failed, continuing",
                    strategy.name,
                    agent.agent_id,
                    exc_info=True,
                )
        return any_acted
